controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java

```diff
@@ -278,11 +285,29 @@ public CommitRequestHandler(final StreamMetadataStore streamMetadataStore,
                     // primary id is taken for creation of txn-segment name and secondary part is erased and replaced with
                     // transaction's epoch.
                     // And we are creating duplicates of txn epoch keeping the primary same.
-                    .thenCompose(v -> streamMetadataTasks.notifyTxnCommit(scope, stream, segments, txnId));
+                    .thenCompose(v -> streamMetadataTasks.notifyTxnCommit(scope, stream, segments, txnId))
+                    .thenCompose(v -> streamMetadataTasks.getSegmentsSize(scope, stream, segments, delegationToken)
+                        .thenCompose(map -> {
+                            // Note: since we are taking the size, in idempotent case we dont necessarily have the exact 
+                            // commit offset for a transcation. 
+                            return transactionCommitCompletionCallback(scope, stream, txnId, map, context);
+                        }));
         }
         return future;
     }
 
+    private CompletableFuture<Void> transactionCommitCompletionCallback(String scope, String stream, UUID txnId,
+                                                                        Map<Long, Long> position, OperationContext context) {
+
+        return streamMetadataStore.getTransactionData(scope, stream, txnId, context, executor)
+                .thenCompose(txnRecord -> {
+                    String writerId = txnRecord.getWriterId();
+                    long commitTime = txnRecord.getCreationTime();
+                    // TODO: shivesh: watermarking: generate a writer mark. We have writer id, time and position
```

Comment:
There are multiple comments related to watermarking. Should we target the watermarking branch?

Reply:
With this branch i have just marked place holders for watermarking. 
This branch only deals with transaction ordering.
The only thing related to watermarking is the `writerId` and `time` fields have also been included in controller metadata and grpc with their default values being sent as of now.